src/H5FDint.c

```diff
@@ -455,13 +476,27 @@ H5FD_read_vector(H5FD_t *file, uint32_t count, H5FD_mem_t types[], haddr_t addrs
                             (unsigned long long)eoa);
         }
     }
+    else
+        /* We must still check if this is a raw data read */
+        for (i = 0; i < count && types[i] != H5FD_MEM_NOLIST; i++)
+            if (types[i] == H5FD_MEM_DRAW) {
+                is_raw = TRUE;
+                break;
+            }
 
     /* if the underlying VFD supports vector read, make the call */
     if (file->cls->read_vector) {
-
         if ((file->cls->read_vector)(file, dxpl_id, count, types, addrs, sizes, bufs) < 0)
-
             HGOTO_ERROR(H5E_VFL, H5E_READERROR, FAIL, "driver read vector request failed");
+
+        /* Set actual selection I/O mode, if this is a raw data operation */
+        if (is_raw) {
```

Comment:
Since the scalar I/O pathway is checking that the raw data I/O is `size > 0`, should we be checking that as well in the vector I/O pathway, or does that matter?

Reply:
So we could do this for vector without too much problem, but to implement it for selection we'd have to check all the dataspaces which could be expensive for the "_id" routines which would have to resolve all of those IDs into H5S_t *s. Vector already avoids modifying the io mode when passes 0 vectors (count == 0),, and that could easily be added to selection. I suppose you could make an argument for reverting this change for scalar (it is making the callback after all), though I'd have to change the test code.

Reply:
I decided to revert the checking for size == 0. There's a fairly length comment explaining some of this behavior in the test now.